GitCommands/Git/GitModule.cs

```diff
@@ -1508,11 +1508,34 @@ public string FormatPatch(string from, string to, string output)
             return result;
         }
 
-        public string Tag(string tagName, string revision, bool annotation, bool force)
-        {
-            if (annotation)
-                return RunGitCmd(string.Format("tag \"{0}\" -a {1} -F \"{2}\\TAGMESSAGE\" -- \"{3}\"", tagName.Trim(), (force ? "-f" : ""), GetGitDirectory(), revision));
-            return RunGitCmd(string.Format("tag {0} \"{1}\" \"{2}\"", (force ? "-f" : ""), tagName.Trim(), revision));
+        public string Tag(string tagName, string revision, bool force, int tagIdx = 0, string keyId = "")
+        {
+            string strCommand = "";
+
+            switch (tagIdx)
+            {
+                /* Annotate */
+                case 1:
+                    strCommand = $"tag \"{tagName.Trim()}\" -a {(force ? " - f" : "")} -F \"{GetGitDirectory()}\\TAGMESSAGE\" -- \"{revision}\"";
+                    break;
+                
+                /* Sign with default GPG */
+                case 2:
+                    strCommand = $"tag \"{tagName.Trim()}\" -s {(force ? " - f" : "")} -F \"{GetGitDirectory()}\\TAGMESSAGE\" -- \"{revision}\"";
+                    break;
+                
+                /* Sign with specific GPG */
+                case 3:
+                    strCommand = $"tag \"{tagName.Trim()}\" -u {keyId} {(force ? " - f" : "")} -F \"{GetGitDirectory()}\\TAGMESSAGE\" -- \"{revision}\"";
+                    break;
+
+                /* Lightweight */
+                default:
+                    strCommand = $"tag {(force ? "-f" : "")} \"{tagName.Trim()}\" \"{revision}\"";
+                    break;
+            }
+
```

Comment:
The basic implementation looks ok, however there few problems with it.

1. All input parameters must be validated. The method is public and you can't control what is being passed in to it.
I don't suppose either `tagName` or `revision` can be null or empty and, if they are, the method must throw `ArgumentNullException`.
`keyId` must be validated for `case 3` and throw if it isn't supplied.

2. There are multiple repeations (e.g. `tagName.Trim()` or `GetGitDirectory()` or `force ? "-f" : ""`) which should be done once.

3. Similar to my suggestion on your other PR, I would extract this functionality in a separate controller (e.g. TagController) and have `IGitModule` passed as a dependency.
Then you can write unit tests which every code path in your method.
For example:

```
        private const string TagName = "bla";
        private const string Revision = "0123456789";
        private const string WorkingDir = "C:\\Working\\Folder";

        [SetUp]
        public void Setup()
        {
            _module = Substitute.For<IGitModule>();
            _module.GetGitDirectory().Returns(x => WorkingDir);

            _controller = new GitTagController(_module);
        }

        [TestCase(true)]
        [TestCase(false)]
        public void Tag_sign_with_default_gpg(bool force)
        {
            var result = _controller.Tag(TagName, Revision, force, 3, "foo");

            _module.Received(1). RunGitCmd($"tag \"{TagName}\" -s {(force ? " - f" : "")} -F \"{WorkingDir}\\TAGMESSAGE\" -- \"{Revision}\""));
        }
```